File: backend/src/main/java/com/bangchat/module/prompt/dto/CharacterCard.java
package com.bangchat.module.prompt.dto;

import java.util.List;

import lombok.AllArgsConstructor;
import lombok.Data;

/**
 * 角色在特定时间线下的完整信息聚合 DTO。
 * 由 PromptBuilder 从 DB 提取并填充，纯粹的数据载体，不包含业务逻辑。
 *
 * 包含三大块：
 * - 基础信息（姓名/学校/乐队/角色）
 * - 性格（基础标签 + 时间线偏移）
 * - 说话语气（基础描述 + 时间线偏移）
 */
@Data
public class CharacterCard {

    // ---- 基础信息 ----
    private String id;
    private String nameCn;
    private String nameJp;
    private String nameEn;
    private String bandName;
    private String school;
    private String role;
    private String instrument;
    private String stageName;
    private String summary;

    // ---- 性格 ----
    private List<String> personalityTraits;
    private String personalityAdjust;

    // ---- 说话语气 ----
    private String speechStyle;
    private String speechAdjust;

    // ---- 补充细节 ----
    private List<String> details;
    private List<AddressingInfo> addressings;

    // ---- 时间线状态 ----
    private String stateDescription;

    /**
     * 将此角色卡渲染为 prompt 中的角色描述段落。
     * 输出格式包含：角色名与身份、学校、乐队与角色、概要、
     * 当前状态、性格标签、说话语气、行为特征、称呼方式。
     *
     * @return 格式化的角色描述文本，可直接嵌入 System Prompt
     */
    public String toPromptSection() {
        StringBuilder sb = new StringBuilder();

        sb.append("【").append(nameCn);
        if (stageName != null && !stageName.isEmpty()) {
            sb.append(" / ").append(stageName);
        }
        sb.append("】（").append(nameJp).append(" / ").append(nameEn).append("）\n");

        sb.append("学校：").append(school).append("\n");
        sb.append("乐队：").append(bandName).append("  |  角色：").append(role);
        if (instrument != null && !instrument.isEmpty()) {
            sb.append("（").append(instrument).append("）");
        }
        sb.append("\n");

        if (summary != null && !summary.isEmpty()) {
            sb.append("概要：").append(summary).append("\n");
        }

        if (stateDescription != null && !stateDescription.isEmpty()) {
            sb.append("当前状态：").append(stateDescription).append("\n");
        }

        sb.append("性格：");
        if (personalityTraits != null && !personalityTraits.isEmpty()) {
            sb.append(String.join("、", personalityTraits));
        }
        if (personalityAdjust != null && !personalityAdjust.isEmpty()) {
            sb.append("（").append(personalityAdjust).append("）");
        }
        sb.append("\n");

        sb.append("说话语气：");
        if (speechStyle != null && !speechStyle.isEmpty()) {
            sb.append(speechStyle);
        }
        if (speechAdjust != null && !speechAdjust.isEmpty()) {
            if (speechStyle != null && !speechStyle.isEmpty()) sb.append("；");
            sb.append(speechAdjust);
        }
        sb.append("\n");

        if (details != null && !details.isEmpty()) {
            sb.append("行为特征：\n");
            for (String d : details) {
                sb.append("  - ").append(d).append("\n");
            }
        }

        if (addressings != null && !addressings.isEmpty()) {
            sb.append("称呼方式：\n");
            for (AddressingInfo a : addressings) {
                sb.append("  - 称呼").append(a.getTargetName())
                  .append("为「").append(a.getAlias()).append("」");
                if (a.getNote() != null && !a.getNote().isEmpty()) {
                    sb.append("（").append(a.getNote()).append("）");
                }
                sb.append("\n");
            }
        }

        return sb.toString();
    }

    @Data
    @AllArgsConstructor
    public static class AddressingInfo {
        private String targetName;
        private String alias;
        private String note;
    }
}

File: bangchat/lib/models/character_card.dart
/// 角色在特定时间线下的完整信息聚合 DTO。
///
/// 由 PromptBuilder 从 DB 提取并填充，纯粹的数据载体。
/// 字段为可变，方便 PromptBuilder 逐步填充各子表数据。
class CharacterCard {
  final String id;
  final String nameCn;
  final String nameJp;
  final String nameEn;
  final String bandName;
  String school;
  String role;
  String instrument;
  String stageName;
  String summary;
  List<String> personalityTraits;
  String personalityAdjust;
  String speechStyle;
  String speechAdjust;
  List<String> details;
  List<AddressingInfo> addressings;
  String stateDescription;

  CharacterCard({
    required this.id,
    required this.nameCn,
    required this.nameJp,
    required this.nameEn,
    required this.bandName,
    this.school = '',
    this.role = '',
    this.instrument = '',
    this.stageName = '',
    this.summary = '',
    List<String>? personalityTraits,
    this.personalityAdjust = '',
    this.speechStyle = '',
    this.speechAdjust = '',
    List<String>? details,
    List<AddressingInfo>? addressings,
    this.stateDescription = '',
  })  : personalityTraits = personalityTraits ?? [],
        details = details ?? [],
        addressings = addressings ?? [];

  /// 将角色卡渲染为 prompt 中的角色描述段落。
  String toPromptSection() {
    final sb = StringBuffer();

    sb.write('【$nameCn');
    if (stageName.isNotEmpty) {
      sb.write(' / $stageName');
    }
    sb.write('】（$nameJp / $nameEn）\n');

    sb.write('学校：$school\n');
    sb.write('乐队：$bandName  |  角色：$role');
    if (instrument.isNotEmpty) {
      sb.write('（$instrument）');
    }
    sb.write('\n');

    if (summary.isNotEmpty) {
      sb.write('概要：$summary\n');
    }

    if (stateDescription.isNotEmpty) {
      sb.write('当前状态：$stateDescription\n');
    }

    sb.write('性格：');
    if (personalityTraits.isNotEmpty) {
      sb.write(personalityTraits.join('、'));
    }
    if (personalityAdjust.isNotEmpty) {
      sb.write('（$personalityAdjust）');
    }
    sb.write('\n');

    sb.write('说话语气：');
    if (speechStyle.isNotEmpty) {
      sb.write(speechStyle);
    }
    if (speechAdjust.isNotEmpty) {
      if (speechStyle.isNotEmpty) sb.write('；');
      sb.write(speechAdjust);
    }
    sb.write('\n');

    if (details.isNotEmpty) {
      sb.write('行为特征：\n');
      for (final d in details) {
        sb.write('  - $d\n');
      }
    }

    if (addressings.isNotEmpty) {
      sb.write('称呼方式：\n');
      for (final a in addressings) {
        sb.write('  - 称呼${a.targetName}为「${a.alias}」');
        if (a.note.isNotEmpty) {
          sb.write('（${a.note}）');
        }
        sb.write('\n');
      }
    }

    return sb.toString();
  }
}

/// 角色称呼信息。
class AddressingInfo {
  final String targetName;
  final String alias;
  final String note;

  const AddressingInfo({
    required this.targetName,
    required this.alias,
    this.note = '',
  });
}
